core/src/main/scala/org/apache/spark/ExecutorAllocationManager.scala

```diff
@@ -829,12 +868,25 @@ private[spark] class ExecutorAllocationManager(
       numTotalTasks - numRunning
     }
 
+    def pendingUnschedulableTasksPerResourceProfile(rp: Int): Int = {
```

Comment:
The method is returning number of task sets which are not schedulable - not number of tasks which are not scheduable.
`maxNumExecutorsNeededPerResourceProfile` is relying on this being task count - we will need to fix it here and in the event being fired.

Reply:
In the blacklisting logic, https://github.com/apache/spark/blob/263f04db865920d9c10251517b00a1b477b58ff1/core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala#L615 - here we just check for the first task which is currently unschedulable (note: there could be more which we would eventually come back to), which is why we will only have 1 unschedulable task per taskset, thats why only we keep track of the unschedulableTaskSets and not the individual tasks in the taskSets themselves. So in short # of unschedulableTaskSets = # of unschedulableTasks fr now.

Reply:
I will refactor the name of this method to make it clear.

Reply:
We should either base additional allocations on:
* Number of unschedulable tasks - this could potentially result in overallocation
* Do something similar to speculative tasks - so that there is atleast one executor allocation to make progress.

Currently, we are using unschedulable tasksets as a proxy for number of unschedulable tasks - which is a hybrid of both of these approaches.

Also note that given the discussion on clearing `unschedulableTaskSets`, we could toggle between asking for more resources and asking for less (as soon as a taskset becomes schedulable).